packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp

```diff
@@ -3940,35 +3839,51 @@ namespace Tpetra {
       /// \param debug [in] If true, read in binary mode.
       static Teuchos::RCP<multivector_type>
       readDenseFile (const std::string& filename,
-                     const Teuchos::RCP<const comm_type>& comm,
+                     const trcp_tcomm_t& comm,
                      Teuchos::RCP<const map_type>& map,
                      const bool tolerant=false,
                      const bool debug=false,
                      const bool binary=false)
       {
-        using Teuchos::broadcast;
-        using Teuchos::outArg;
 
+        std::ifstream in = Reader::openFile(comm, filename, true, binary ? std::ios::binary : std::ios::in);
+
+        return readDense (in, comm, map, tolerant, debug, binary);
+      }
+
+      /**
+       * @brief Open an input file stream safely on the root rank (0).
+       *
+       * Only open the file on root process. Test carefully to make
+       * sure that the file opened successfully and broadcast that
+       * result to all processes to prevent a hang on exception
+       * throw.
+       */
+      static std::ifstream openFile(
+        const trcp_tcomm_t& comm,
+        const std::string& filename, const bool safe = true,
+        std::ios_base::openmode mode = std::ios_base::in
+      ){
```

Comment:
How about `openFileOnRankZero`?

Let's document what properties the returned `ifstream` has for the other ranks. Alternatively, return an `std::optional<std::ifstream>`

Reply:
* I think using `std::optional` would lead to many additional changes, especially where the underlying value of the `std::optional<std::ifstream>` will be used... If you agree, I'll just update the documentation.
* Ok for renaming to `openFileOnRankZero`

Reply:
I am happy with the path of least resistance 😄